users/dierkes/preprocessing/segment_cutting.py

```diff
@@ -0,0 +1,107 @@
+import pathlib
+import soundfile as sf
+import numpy as np
+import json
+import multiprocessing
+import argparse
+import tqdm
+import os
+import shutil
+import logging
+
+from sisyphus import Job, Task, tk
+
+class CutSpeechSegmentsFromAudio(Job):
+    """
+    This Job uses a transcription file to cut all segments containing speech from the audio files
+    and stitches them together to files of a target length. Segments that are
+    longer than the target_length are stored as individual files.
+
+
+    """
+    def __init__(self, audio_dir, segments_dir, target_length=20, n_workers=32, file_extension='.wav'):
+        """
+
+        :param audio_dir: path to all the audio files. Subfolders are not yet supported in audio
+            directory
+        :param segments_dir: path to the transcriptions. Subfolders are not yet supported in
+            transcription directory
+        :param target_length: target length of the stitched extracted speech
+        :param n_workers:
+        :param file_extension: currently supported are wav, flac and mp3
+        """
+        self.audio_dir = audio_dir
+        self.segments_dir = segments_dir
+        self.target_length = target_length
+        self.n_workers = n_workers
+        self.file_extension = file_extension
+        assert file_extension in ['.wav', '.flac', '.mp3']
+
+        self.out_audio_folder = self.output_path("audio/", directory=True)
+        self.cut_rqmt = {'time': 8, 'mem': 8, 'cpu': 4}
+
+    def tasks(self):
+        yield Task("cut_speech_segments", rqmt=self.cut_rqmt, mini_task=True)
+
+    def cut_file(self, task):
+        path_file, root_out, target_len_sec, extension, segments = task
+        file_name = path_file.stem
+        meta_file_path = f"{segments}/{file_name}.json"
+
+        with open(meta_file_path, 'r') as f:
+            meta = json.loads(f.read())["value"]["segments"]
+
+        to_cut = []
+        for seg in meta:
+            to_cut += [(seg["start"], seg["end"])]
+
+        data, samplerate = sf.read(path_file)
+        assert len(data.shape) == 1
+        assert samplerate == 8000
+
+        to_stitch = []
+        length_accumulated = 0.0
+
+        i = 0
+        for start, end in to_cut:
+            start_index = int(start * samplerate)
+            end_index = int(end * samplerate)
+            slice = data[start_index:end_index]
+
+            # if a slice is longer than target_len_sec, we put it entirely in it's own piece
+            if length_accumulated + (end - start) > target_len_sec and length_accumulated > 0:
```

Comment:
With this logic of dumping everything if we would exceed the target length, we might still get short segments if a very long segment follows. Maybe this is not important, just as a note.

Reply:
this is meant to be this way so that segments longer than the target_length will always be in an individual file

Reply:
It is actually meant to be this way so that segments longer than the target_length will always be in an individual file

Reply:
But they will only be in an individual file, if the buffer `to_stitch` is empty before, right?

Reply:
Yes, but it will always be empty because we empty it before adding the new segment. If a segment exceeds the `target_length` we first store everything in `to_stitch` in a file and empty `to_stitch`. Then the new segmented is added to `to_stitch`. Now in case this segment is already larger than the `target_length` it will be stored in its own file in the next iteration of the loop. So segments longer than the `target_length` will always be in their own file.

Reply:
Ah yes, I see.